### ISSUE TITLE
Modify shape layer data programmatically (add/modify shapes), unclear how to

### ISSUE THREAD

**Issue by VolkerH** [napari/napari#2784]
Adding this issue to keep track of some unclear points about adding data to shape layers programmatically.
This is from  https://napari.zulipchat.com/#narrow/stream/212875-general/topic/adding.20shape.20to.20shape.20layer.20programmatically

I was  trying to add some shapes to an existing shape layer.

My naive attempt was to appending to the shape layers `.data`
```
# sl is a shape layer
sl.data.append(np.array([[...][...]])
```
This does not work  as  `.data` returns a shallow copy of the list, so the array is not actually appended to the layer data.
@tlambert03 suggested that the `.data` property should probably return a tuple to indicate that this is immutable.

As it is a shallow copy it is possible to modify the data of indivdual shapes like this, though:
```
# overwrite an individual shape coordinate with a different value
sl.data[0][0,0] = 42.0
```
This change does indeed modify the underlying shape data, however I could not force a redraw using `sl.refresh()`.

I then found out that what I primarily want to do (add a shape) is possible via
```
sl._add_shapes(np.array(([...]))
```
However, the underscore indicates that this method is private and should not be relied on by users.

These are just some individual observations, but the general picture seems to be that there is a lack of documentation on how users can programmatically change data in existing shape layers.

**Comment by sofroniewn** [comment]
Yeah we should clean this up, make it easier to understand

> This does not work as .data returns a shallow copy of the list, so the array is not actually appended to the layer data.
@tlambert03 suggested that the .data property should probably return a tuple to indicate that this is immutable.

Indeed I think we should make this a tuple, otherwise we could make it an evented list and hook everything up with events, which would be nice too

> This change does indeed modify the underlying shape data, however I could not force a redraw using sl.refresh().

Hmm this surprises me, we'll have to look

> However, the underscore indicates that this method is private and should not be relied on by users.

There is also a public `layer.add` method, https://napari.org/api/stable/napari.layers.Shapes.html#napari.layers.Shapes.add
but maybe we need to document that more clearly/ make it more discoverable. I think it should do what you want though

**Comment by psobolewskiPhD** [comment]
@VolkerH @DragaDoncila 
Looking at this issue and the docs at:
https://napari.org/stable/howtos/layers/shapes.html#adding-new-shapes
it seems like there is documentation for this outside the API docs.
I tested using that code and as long as `shape_type` is properly passed, everything seems to work.
What do y'all think? Is this closable? Or do we need to make an issue in [napari/docs](https://github.com/napari/docs) to address this?

**Comment by DragaDoncila** [comment]
@psobolewskiPhD I think it kind of is surprising that editing `shapes_layer.data` doesn't lead to an update in the layer, and can't be triggered by refresh. Having looked a little bit into why this is, I think the best path forward here if we don't want to do major surgery on the shapes layer is to at a minimum, update the shapes tutorial to mention `.data` as a **read-only** copy of the data, and explicitly state that updates to the shapes should be made exclusively using `add_shape`, `remove_shape` and vertex functions

**Comment by psobolewskiPhD** [comment]
Sounds good. I'll make an issue on the docs repo to flag this and add an admonition regarding `.data` and `.append`.
Then once we update the docs we can decide if that's a sufficient fix or not.

**Comment by DragaDoncila** [comment]
Thanks @psobolewskiPhD. I think it would also be good to make the return type of `shapes.data` a `tuple` and update the API docs to reflect that. Maybe we should update the title of this issue to ask for that and mark it `good first issue` and `ready for work`? Or we could close this issue and open another smaller more specific one idk what's better.

**Comment by psobolewskiPhD** [comment]
What about just making a new issue that for that API change that references this and the docs issue and then closing everything once the two PRs land?

**Comment by DragaDoncila** [comment]
That seems fine too!
